File: src/main/java/org/lanternpowered/server/network/vanilla/advancement/NetworkAdvancement.java
/*
 * Lantern
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) SpongePowered <https://www.spongepowered.org>
 * Copyright (c) contributors
 *
 * This work is licensed under the terms of the MIT License (MIT). For
 * a copy, see 'LICENSE.txt' or <https://opensource.org/licenses/MIT>.
 */
package org.lanternpowered.server.network.vanilla.advancement;

import com.google.common.base.MoreObjects;
import com.google.common.collect.Iterables;
import org.lanternpowered.server.network.buffer.ByteBuffer;
import org.lanternpowered.server.network.message.codec.CodecContext;

import java.util.Collection;

import org.checkerframework.checker.nullness.qual.Nullable;

public final class NetworkAdvancement {

    private final String id;
    @Nullable private final String parentId;
    @Nullable private final NetworkAdvancementDisplay display;
    private final Collection<String> criteria;
    private final String[][] requirements;

    public NetworkAdvancement(String id, @Nullable String parentId, @Nullable NetworkAdvancementDisplay display,
            Collection<String> criteria, String[][] requirements) {
        this.requirements = requirements;
        this.parentId = parentId;
        this.criteria = criteria;
        this.display = display;
        this.id = id;
    }

    void write(CodecContext ctx, ByteBuffer buf) {
        buf.writeString(this.id);
        buf.writeBoolean(this.parentId != null);
        if (this.parentId != null) {
            buf.writeString(this.parentId);
        }
        buf.writeBoolean(this.display != null);
        if (this.display != null) {
            this.display.write(ctx, buf);
        }
        buf.writeVarInt(this.criteria.size());
        this.criteria.forEach(buf::writeString);
        buf.writeVarInt(this.requirements.length);
        for (String[] requirements : this.requirements) {
            buf.writeVarInt(requirements.length);
            for (String requirement : requirements) {
                buf.writeString(requirement);
            }
        }
    }

    @Override
    public String toString() {
        return MoreObjects.toStringHelper(this)
                .omitNullValues()
                .add("id", this.id)
                .add("parentId", this.parentId)
                .add("display", this.display)
                .add("criteria", Iterables.toString(this.criteria))
                .toString();
    }
}

File: src/main/kotlin/org/lanternpowered/server/network/vanilla/advancement/NetworkAdvancement.kt
/*
 * Lantern
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) SpongePowered <https://www.spongepowered.org>
 * Copyright (c) contributors
 *
 * This work is licensed under the terms of the MIT License (MIT). For
 * a copy, see 'LICENSE.txt' or <https://opensource.org/licenses/MIT>.
 */
package org.lanternpowered.server.network.vanilla.advancement

import org.lanternpowered.api.util.ToStringHelper
import org.lanternpowered.api.util.collections.contentToString
import org.lanternpowered.server.network.buffer.ByteBuffer
import org.lanternpowered.server.network.value.ContextualValueCodec
import org.lanternpowered.server.network.packet.codec.CodecContext

class NetworkAdvancement(
        private val id: String,
        private val parentId: String?,
        private val display: NetworkAdvancementDisplay?,
        private val criteria: Collection<String>,
        private val requirements: Array<Array<String>>
) {

    companion object : ContextualValueCodec<NetworkAdvancement> {

        override fun write(ctx: CodecContext, buf: ByteBuffer, value: NetworkAdvancement) {
            buf.writeString(value.id)
            buf.writeBoolean(value.parentId != null)
            if (value.parentId != null)
                buf.writeString(value.parentId)
            buf.writeBoolean(value.display != null)
            if (value.display != null)
                NetworkAdvancementDisplay.write(ctx, buf, value.display)
            buf.writeVarInt(value.criteria.size)
            for (criterion in value.criteria)
                buf.writeString(criterion)
            buf.writeVarInt(value.requirements.size)
            for (requirements in value.requirements) {
                buf.writeVarInt(requirements.size)
                for (requirement in requirements)
                    buf.writeString(requirement)
            }
        }

        override fun read(ctx: CodecContext, buf: ByteBuffer): NetworkAdvancement = throw UnsupportedOperationException()
    }

    override fun toString(): String = ToStringHelper(this)
            .omitNullValues()
            .add("id", this.id)
            .add("parentId", this.parentId)
            .add("display", this.display)
            .add("criteria", this.criteria.contentToString())
            .toString()
}
